pkg/controller/discovery/auth/rbac.go

```diff
@@ -0,0 +1,324 @@
+			allowed, err := r.Allow(request)
+			if err != nil {
+				Log.Trace(err)
+				return false, err
+			}
+			if !allowed {
+				return false, nil
+			}
+		}
+	}
+
+	return mAllowed, nil
+}
+
+//
+// Get whether authenticated
+func (r *RBAC) Authenticated() bool {
+	return r != nil && *r.authenticated
+}
+
+//
+// Get the user name for the token.
+func (r *RBAC) User() string {
+	if r.Authenticated() {
+		return r.tokenStatus.User.Username
+	}
+
+	return ""
+}
+
+//
+// Get the user ID for the token.
+func (r *RBAC) UID() string {
+	if r.Authenticated() {
+		return r.tokenStatus.User.UID
+	}
+
+	return ""
+}
+
+//
+// Get the user's groups.
+func (r *RBAC) Groups() []string {
+	if r.Authenticated() {
+		return r.tokenStatus.User.Groups
+	}
+
+	return []string{}
+}
+
+//
+// Get the extra fields for the token.
+func (r *RBAC) Extra() map[string]v1.ExtraValue {
+	extra := map[string]v1.ExtraValue{}
+	if r.Authenticated() {
+		for k, v := range r.tokenStatus.User.Extra {
+			extra[k] = append(
+				v1.ExtraValue{},
+				v...)
+		}
+	}
+
+	return extra
+}
+
+//
+// Do subject access review.
+func (r *RBAC) sar(request *Review) (bool, error) {
+	sar := v1.SubjectAccessReview{
+		Spec: v1.SubjectAccessReviewSpec{
+			ResourceAttributes: request,
+			Groups:             r.Groups(),
+			User:               r.User(),
+			UID:                r.UID(),
+			Extra:              r.Extra(),
+		},
+	}
+	err := r.Client.Create(context.TODO(), &sar)
+	if err != nil {
+		Log.Trace(err)
+		return false, err
+	}
+
+	return sar.Status.Allowed, nil
+}
+
+//
+// Authenticate the token.
+// Build the `subjectClient`.
+func (r *RBAC) Authenticate() (bool, error) {
+	if r.authenticated != nil {
+		return *r.authenticated, nil
+	}
+	r.authenticated = pointer.BoolPtr(false)
+	tr := &auth.TokenReview{
+		Spec: auth.TokenReviewSpec{
+			Token: r.Token,
+		},
+	}
+	err := r.Client.Create(context.TODO(), tr)
+	if err != nil {
+		Log.Trace(err)
+		return false, err
+	}
+	if !tr.Status.Authenticated {
+		return false, err
+	}
+	r.tokenStatus = tr.Status
+	r.authenticated = pointer.BoolPtr(true)
+
+	return true, nil
+}
+
+//
+// Provides project-based authorization.
+type PBAC struct {
+	// Token
+	Token string
+	// Cluster
+	Cluster *migapi.MigCluster
+	// Project API supported by the cluster.
+	supported bool
+	// List of projects.
+	projects map[string]bool
+}
+
+//
+// Project API is supported by the cluster.
+func (p *PBAC) Supported() bool {
+	p.Load()
+	return p.supported
+}
+
+//
+// The token is allowed access to the specified namespace.
+func (p *PBAC) Allow(ns string) bool {
+	p.Load()
+	_, found := p.projects[ns]
+	return found
+}
+
+//
+// Load projects (as needed).
+func (p *PBAC) Load() {
+	if p.projects != nil {
+		return
+	}
+	p.projects = map[string]bool{}
+	tokenClient, err := p.client()
+	if err != nil {
+		return
+	}
+	list := &project.ProjectList{}
+	err = tokenClient.List(context.TODO(), nil, list)
+	if err != nil {
+		return
+	}
+	p.supported = true
```

Comment:
What happens when the API server responds with a transient error? looks like there in that case the user will no use project API even if the project is available but apiserver returned and expected error? 

This does not feel like the right thing to do, it feels like figuring out if projects API is supported or not using discovery API is a safer alternative.

Reply:
After the discussion, it was concluded that since RBAC and PBAC structs are instantiated on every request, handling transient api server will not break functionality, it will only fall back to less optimized SAR checks. Resolving because this does not need to change

Reply:
@alaypatel07 what is PBAC?